File: lawhook-python/lawhook/exceptions.py
# ─────────────────────────────────────────────────
#  Lawhook SDK — Exception Hierarchy
#
#  All SDK errors inherit from LawhookError so
#  developers can catch broadly or specifically:
#
#      try:
#          client.changes.list()
#      except LawhookError as e:
#          print(e)
#
#      try:
#          client.changes.get("bad_id")
#      except NotFoundError:
#          print("Change not found")
# ─────────────────────────────────────────────────

class LawhookError(Exception):
    """
    Base exception for all Lawhook SDK errors.
 
    Attributes:
        message:     human-readable error message
        status_code: HTTP status code, if applicable
        response:    raw response body (dict), if available
    """
    
    def __init__(
        self,
        message: str,
        status_code: int | None = None,
        response: dict | None = None,
    ):
        self.message = message
        self.status_code = status_code
        self.response = response
        super().__init__(message)
        
    def __str__(self) -> str:
        if self.status_code:
            return f"{self.status_code} Error: {self.message}"
        return self.message
    
class AuthenticationError(LawhookError):
    """
    Raised on HTTP 401.
 
    Indicates the API key is missing, invalid, or revoked.
    """
    pass

class NotFoundError(LawhookError):
    """
    Raised on HTTP 404.
 
    The requested resource (subscription, change) does not
    exist or does not belong to your API key.
    """    
    pass

class ValidationError(LawhookError): 
    """
    Raised on HTTP 422.
 
    Request data failed validation — check the `response`
    attribute for field-level details from the API.
    """
    pass

class RateLimitError(LawhookError):
    """
    Raised on HTTP 429.
 
    Attributes:
        retry_after: seconds to wait before retrying,
                     read from the Retry-After header.
                     None if the header was missing.
 
    Example:
        try:
            client.changes.list()
        except RateLimitError as e:
            time.sleep(e.retry_after or 60)
            client.changes.list()
    """
    
    def __init__(
        self,
        message: str,
        status_code: int | None = None,
        response: dict | None = None,
        retry_after: int | None = None,
    ):
        self.retry_after = retry_after
        super().__init__(message, status_code, response)
        
class ServerError(LawhookError):
    """
    Raised on HTTP 5xx.
 
    Indicates an error on Lawhook's side. Safe to retry
    with exponential backoff.
    """
    pass

class ConnectionError(LawhookError):
    """
    Raised when the request could not be sent at all —
    network failure, DNS failure, or timeout.
 
    Distinct from ServerError: the request never reached
    Lawhook's servers.
    """
    pass

def raise_for_status(status_code: int, body: dict | None) -> None:
    """
    Inspect an HTTP status code and raise the appropriate
    LawhookError subclass. Called internally by LawhookClient
    after every request — SDK users never call this directly.
 
    Does nothing if status_code indicates success (2xx).
    """
    if 200 <= status_code < 300:
        return
 
    detail = "Unknown error"
    if isinstance(body, dict):
        detail = body.get("detail", detail)
 
    if status_code == 401:
        raise AuthenticationError(
            detail or "Invalid or missing API key",
            status_code=status_code,
            response=body,
        )
 
    if status_code == 404:
        raise NotFoundError(
            detail or "Resource not found",
            status_code=status_code,
            response=body,
        )
 
    if status_code == 422:
        raise ValidationError(
            detail or "Validation failed",
            status_code=status_code,
            response=body,
        )
 
    if status_code == 429:
        # retry_after is attached separately by the client
        # since it comes from a response header, not the body
        raise RateLimitError(
            detail or "Rate limit exceeded",
            status_code=status_code,
            response=body,
        )
 
    if status_code >= 500:
        raise ServerError(
            detail or f"Lawhook server error ({status_code})",
            status_code=status_code,
            response=body,
        )
 
    # Catch-all for any other 4xx
    raise LawhookError(
        detail or f"Request failed ({status_code})",
        status_code=status_code,
        response=body,
    )

File: lawhook-typescript/src/exceptions.ts
// src/exceptions.ts
// ─────────────────────────────────────────────────
//  Lawhook SDK — Exception Hierarchy
//
//  All SDK errors extend LawhookError so callers
//  can catch broadly or specifically:
//
//      try {
//          await client.changes.list()
//      } catch (err) {
//          if (err instanceof RateLimitError) {
//              await sleep(err.retryAfter * 1000)
//          } else if (err instanceof LawhookError) {
//              console.error(err.message)
//          }
//      }
// ─────────────────────────────────────────────────

export class LawhookError extends Error {
    /** HTTP status code, if applicable */
    statusCode?: number
    /** Raw response body from the API, if available */
    response?: Record<string, unknown>

    constructor(
        message: string,
        statusCode?: number,
        response?: Record<string, unknown>,
    ) {
        super(message)
        this.name = "LawhookError"
        this.statusCode = statusCode
        this.response = response
        // Restore prototype chain (required for extending Error in TS)
        Object.setPrototypeOf(this, new.target.prototype)
    }
}

/** Raised on HTTP 401 — invalid or missing API key */
export class AuthenticationError extends LawhookError {
    constructor(message: string, response?: Record<string, unknown>) {
        super(message, 401, response)
        this.name = "AuthenticationError"
        Object.setPrototypeOf(this, new.target.prototype)
    }
}

/** Raised on HTTP 404 — resource doesn't exist or isn't yours */
export class NotFoundError extends LawhookError {
    constructor(message: string, response?: Record<string, unknown>) {
        super(message, 404, response)
        this.name = "NotFoundError"
        Object.setPrototypeOf(this, new.target.prototype)
    }
}

/** Raised on HTTP 422 — request data failed validation */
export class ValidationError extends LawhookError {
    constructor(message: string, response?: Record<string, unknown>) {
        super(message, 422, response)
        this.name = "ValidationError"
        Object.setPrototypeOf(this, new.target.prototype)
    }
}

/**
 * Raised on HTTP 429 — rate limit exceeded.
 *
 * @example
 * try {
 *     await client.changes.list()
 * } catch (err) {
 *     if (err instanceof RateLimitError) {
 *         await new Promise(r => setTimeout(r, (err.retryAfter ?? 60) * 1000))
 *     }
 * }
 */
export class RateLimitError extends LawhookError {
    /** Seconds to wait before retrying, from Retry-After header */
    retryAfter: number | null

    constructor(
        message: string,
        retryAfter: number | null = null,
        response?: Record<string, unknown>,
    ) {
        super(message, 429, response)
        this.name = "RateLimitError"
        this.retryAfter = retryAfter
        Object.setPrototypeOf(this, new.target.prototype)
    }
}

/** Raised on HTTP 5xx — error on Lawhook's side, safe to retry */
export class ServerError extends LawhookError {
    constructor(
        message: string,
        statusCode: number,
        response?: Record<string, unknown>,
    ) {
        super(message, statusCode, response)
        this.name = "ServerError"
        Object.setPrototypeOf(this, new.target.prototype)
    }
}

/**
 * Raised when the request couldn't be sent at all —
 * network failure, DNS failure, or timeout.
 * Distinct from ServerError: the request never reached
 * Lawhook's servers.
 */
export class ConnectionError extends LawhookError {
    constructor(message: string) {
        super(message)
        this.name = "ConnectionError"
        Object.setPrototypeOf(this, new.target.prototype)
    }
}

/**
 * Inspect an HTTP status code and throw the appropriate
 * LawhookError subclass. Called internally after every
 * request — SDK users never call this directly.
 * Does nothing if status is 2xx.
 */
export function throwForStatus(
    status: number,
    body: Record<string, unknown> | null,
    retryAfter: number | null = null,
): void {
    if (status >= 200 && status < 300) return

    const detail =
        typeof body?.detail === "string" ? body.detail : "Unknown error"

    if (status === 401) throw new AuthenticationError(detail, body ?? undefined)
    if (status === 404) throw new NotFoundError(detail, body ?? undefined)
    if (status === 422) throw new ValidationError(detail, body ?? undefined)
    if (status === 429) {
        throw new RateLimitError(detail, retryAfter, body ?? undefined)
    }
    if (status >= 500) throw new ServerError(detail, status, body ?? undefined)

    throw new LawhookError(detail, status, body ?? undefined)
}